presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java

```diff
@@ -194,41 +206,83 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
         }
     }
 
-    @ThreadSafe
+    /**
+     * Each LookupSource has it's own copy of OuterPositionTracker instance.
+     * Each of those OuterPositionTracker must be committed after last write
+     * and before first read.
+     *
+     * All instances share visitedPositions array, but it is safe because each thread
+     * starts with visitedPositions filled with false values and marks only some positions
+     * to true. Since we don't care what will be the order of those writes to
+     * visitedPositions, writes can be without synchronization.
+     *
+     * Synchronization in commit() is required because each writing thread must have
```

Comment:
@martint dropping those synchronizations would indeed work just as fine, as long as we keep this `AtomicLong referenceCount` variable. However it is an implementation detail whether `referenceCount` exists or not. In this case. It is there only for the sanity check done later:

```verify(referenceCount.get() == 0);```

We could delete `referenceCount` and code would work just as fine. By dropping synchronization in `commit` and `getVisitedPositions` I would have to add comment to `referenceCount` variable: "DO NOT REMOVE IT, because it's synchronization is required to guarantee memory visibility between `commit` and `getVisitedPositions`". I think that would be less explicit and more hacky compared to `synchronized` methods. And it wouldn't give any performance improvement.